Adding ephemeral nodes to Monsoon

Sadi-a opened the pull request:
Until now, Typhoon needed Flatcar Container Linux to be installed on a machine in order to start working on it. This impedes the implementation of ephemeral workers, hence the reason for this change. This commit allows one to set a variable `enable_install` which defaults to true, and boots the machines live without the need of installing Flatcar Container Linux while using a persistent partition for things such as SSH host keys or Kubernetes certificates. The disk used for this persistent partition will either be a disk specified with its path in the controller or worker's definition or the smallest disk available on a given machine as we do not need to store a lot of information. This is done by introducing conditions and logic  in the yaml templates and sending the controller/worker ignition files instead of the install ignition file as well as making some changes to what is done through SSH, transferring it to the ignition file.

This means that, if a worker were to become unhealthy, after investigating the cause of its failure, it can simply be rebooted with an ipmi `power reset` command or a `systemctl reboot` and it will be reimaged and reprovisioned with the ignition file, except for some of the persistent data such as some Kubernetes certificates or manifests or SSH host keys.

This Pull Request only affects Monsoon's Flatcar Container Linux and does not change Monsoon's Fedora CoreOS.

This pull request has been tested with the test-suite : bootstrapping works as expected and a worker that is restarted registers in the cluster again after finishing its reboot.

Sadi-a committed:
Adding ephemeral worker nodes to Monsoon

Until now, Typhoon needed Flatcar Linux to be installed on a machine
in order to start working on it. This impedes the implementation of
ephemeral workers, hence the reason for this change. This commit
allows one to set a variable `enable_install` which defaults to true,
and boots the machines live without the need of installing Flatcar
Linux while using a persistent partition for things such as SSH host
keys or Kubernetes certificates. The disk used for this persistent
partition will either be a disk specified with its path in the
controller or worker's definition or the smallest disk available on a
given machine as we do not need to store a lot of information.
This is done by introducing conditions and logic  in the yaml
templates and sending the controller/worker ignition files instead of
the install ignition file as well as making some changes to what is
done through SSH, transferring it to the ignition file.

This means that, if a worker were to become unhealthy, after
investigating the cause of its failure, it can simply be rebooted with
an ipmi `power reset` command or a `systemctl reboot` and it will be
reimaged and reprovisioned with the ignition file, except for some
of the persistent data such as some Kubernetes certificates or
manifests or SSH host keys.

Snaipe commented on flatcar-linux/kubernetes/butane/controller.yaml:
```diff
@@ -117,27 +121,70 @@ systemd:
+        DefaultDependencies=no
+        [Mount]
+        What=/dev/disk/by-label/persistent
+        Where=/mnt/persist
```
Do not use /mnt; /mnt is reserved for administrative use when mounting disks for temporary operations. You can create a /persist directory at the root if you need to.

Snaipe commented on flatcar-linux/kubernetes/butane/controller.yaml:
```diff
@@ -117,27 +121,70 @@ systemd:
+        Options=defaults
+        [Install]
+        WantedBy=multi-user.target
+    - name: persistent_preparation.service
+      enabled: true
+      contents: |
+        [Unit]
+        Description= A service used to prepare the persistent disk by creating the required partition if it doesn't exist
+        [Service]
+        Type= oneshot
+        ExecStart=/opt/create_persistent_partition
+        RemainAfterExit= false
+        [Install]
+        WantedBy=multi-user.target
```
Isn't all this already supported as a first-class construct in ignition?

Sadi-a replied:
I thought I would have issues with the `device` part, since what I'm doing here is either using the persist_disk variable if it is specified or choosing the smallest disk available on the machine. 

if we consider `persist_disk` a mandatory variable, I think we can easily switch to this one

Sadi-a replied:
done

Snaipe commented on flatcar-linux/kubernetes/worker/ssh.tf:
```diff
@@ -19,9 +19,11 @@ resource "null_resource" "copy-worker-secrets" {
-      "sudo mv /home/core/kubeconfig /etc/kubernetes/kubeconfig",
-    ]
-  }
+  # provisioner "remote-exec" {
+  #   inline = [
+  #     var.enable_install ?
+  #       "sudo mv /home/core/kubeconfig /etc/kubernetes/kubeconfig" :
+  #       "sudo mv /home/core/kubeconfig /mnt/persist/etc/kubernetes/kubeconfig",
+  #   ]
+  # }
```
Why is this commented out?

Snaipe commented on flatcar-linux/kubernetes/profiles.tf:
```diff
@@ -90,6 +89,10 @@ data "ct_config" "controllers" {
     cluster_domain_suffix  = var.cluster_domain_suffix
     ssh_authorized_key     = var.ssh_authorized_key
+    enable_install         = var.enable_install
+    live_disk              = var.controllers.*.live_disk[count.index]
```
Can we name this persist_disk instead?

Sadi-a commented:
Had to work on this a little, there was an issue where etcd data, which is what is used to store Kubernetes objects in our case, was not persistent, and we would thus lose all our secrets, namespaces and so on when the controller reboots.
Finished fixing this and testing the fix

Sadi-a committed:
Live : make live nodes able to survive a second terraform apply

Formerly, the kubernetes manifests, such as kubeconfig, stored in the
persistent partitions of our nodes would stay even if we launched
terraform apply and rebooted our nodes to take the new configuration.
We would thus have a new kubeconfig file to use and given to a node
that may have been added, while the other nodes would keep the old
kubeconfig.
This commit fixes that by making sure we can overwrite the old
kubernetes manifests when we reboot after a terraform apply.
